common/clock/ratelimiter.go

```diff
@@ -0,0 +1,326 @@
+// The MIT License (MIT)
+
+// Copyright (c) 2017-2020 Uber Technologies Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package clock
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"golang.org/x/time/rate"
+)
+
+type (
+	// Ratelimiter is a test-friendly version of [golang.org/x/time/rate.Limiter],
+	// which can be backed by any TimeSource.  The changes summarize as:
+	//   - Wait has been removed, as it is difficult to precisely mimic, though an
+	//     approximation is likely good enough. However, our Wait-using ratelimiters
+	//     *only* use Wait, and they can have a drastically simplified API and
+	//     implementation if they are handled separately.
+	//   - All MethodAt APIs have been excluded as we do not use them, and they would
+	//     have to bypass the contained TimeSource, which seems error-prone.
+	//   - All `MethodN` APIs have been excluded because we do not use them, but they
```

Comment:
side not: these would be useful if we decide to have different weights per call or its content

Reply:
their interaction with burst is a bit funky and I'm not sure that's something we want to play around with (as opposed to just repeatedly calling `Allow()`)... but yea.  easy and has useful scenarios, I'm pretty confident there won't be any surprises if we add it.

Reply:
They can be useful, if the limit is not based on QPS but BPS.

Reply:
do we have any of those?  I can add them, I just haven't seen it